File: tests/engine/input-processor.test.ts
import { describe, it, expect } from "vitest";
import {
  createInputState,
  processChar,
  processSpace,
  processBackspace,
  processCtrlBackspace,
  processKeystroke,
} from "../../src/engine/input-processor.js";

describe("createInputState", () => {
  it("initializes with correct defaults", () => {
    const state = createInputState(["hello", "world"]);
    expect(state.words).toHaveLength(2);
    expect(state.currentWordIndex).toBe(0);
    expect(state.cursorPosition).toBe(0);
    expect(state.finished).toBe(false);
    expect(state.words[0].word).toBe("hello");
    expect(state.words[0].typed).toBe("");
    expect(state.words[0].completed).toBe(false);
    expect(state.historicalErrorChars).toBe(0);
  });
});

describe("processChar", () => {
  it("adds a correct character", () => {
    const state = createInputState(["hello"]);
    const next = processChar(state, "h");
    expect(next.words[0].typed).toBe("h");
    expect(next.words[0].chars).toHaveLength(1);
    expect(next.words[0].chars[0].correct).toBe(true);
    expect(next.words[0].chars[0].extra).toBe(false);
    expect(next.cursorPosition).toBe(1);
  });

  it("adds an incorrect character", () => {
    const state = createInputState(["hello"]);
    const next = processChar(state, "x");
    expect(next.words[0].typed).toBe("x");
    expect(next.words[0].chars[0].correct).toBe(false);
    expect(next.words[0].chars[0].extra).toBe(false);
    expect(next.historicalErrorChars).toBe(1);
  });

  it("marks extra characters beyond word length", () => {
    let state = createInputState(["hi"]);
    state = processChar(state, "h");
    state = processChar(state, "i");
    state = processChar(state, "x"); // extra
    expect(state.words[0].chars[2].extra).toBe(true);
    expect(state.words[0].chars[2].correct).toBe(false);
    expect(state.cursorPosition).toBe(3);
    expect(state.historicalErrorChars).toBe(1);
  });

  it("does nothing when finished", () => {
    const state = createInputState(["a"]);
    const finished = { ...state, finished: true };
    const next = processChar(finished, "a");
    expect(next.words[0].typed).toBe("");
  });
});

describe("processSpace", () => {
  it("advances to the next word", () => {
    let state = createInputState(["hello", "world"]);
    state = processChar(state, "h");
    state = processChar(state, "e");
    state = processChar(state, "l");
    state = processChar(state, "l");
    state = processChar(state, "o");
    state = processSpace(state);
    expect(state.words[0].completed).toBe(true);
    expect(state.currentWordIndex).toBe(1);
    expect(state.cursorPosition).toBe(0);
  });

  it("does not advance if nothing typed", () => {
    const state = createInputState(["hello", "world"]);
    const next = processSpace(state);
    expect(next.currentWordIndex).toBe(0);
  });

  it("sets finished on last word", () => {
    let state = createInputState(["hi"]);
    state = processChar(state, "h");
    state = processChar(state, "i");
    state = processSpace(state);
    expect(state.finished).toBe(true);
    expect(state.words[0].completed).toBe(true);
  });

  it("allows advancing with incomplete word", () => {
    let state = createInputState(["hello", "world"]);
    state = processChar(state, "h");
    state = processSpace(state);
    expect(state.words[0].completed).toBe(true);
    expect(state.currentWordIndex).toBe(1);
  });
});

describe("processBackspace", () => {
  it("removes the last character", () => {
    let state = createInputState(["hello"]);
    state = processChar(state, "h");
    state = processChar(state, "e");
    state = processBackspace(state);
    expect(state.words[0].typed).toBe("h");
    expect(state.words[0].chars).toHaveLength(1);
    expect(state.cursorPosition).toBe(1);
  });

  it("goes back to previous completed word", () => {
    let state = createInputState(["hi", "there"]);
    state = processChar(state, "h");
    state = processChar(state, "i");
    state = processSpace(state);
    expect(state.currentWordIndex).toBe(1);

    state = processBackspace(state);
    expect(state.currentWordIndex).toBe(0);
    expect(state.words[0].completed).toBe(false);
    expect(state.cursorPosition).toBe(2);
  });

  it("does nothing at start of first word", () => {
    const state = createInputState(["hello"]);
    const next = processBackspace(state);
    expect(next).toEqual(state);
  });

  it("keeps historical errors after deleting an incorrect character", () => {
    let state = createInputState(["hello"]);
    state = processChar(state, "x"); // incorrect
    expect(state.historicalErrorChars).toBe(1);

    state = processBackspace(state);
    expect(state.words[0].typed).toBe("");
    expect(state.words[0].chars).toHaveLength(0);
    expect(state.historicalErrorChars).toBe(1);
  });
});

describe("processCtrlBackspace", () => {
  it("clears the entire current word", () => {
    let state = createInputState(["hello"]);
    state = processChar(state, "h");
    state = processChar(state, "e");
    state = processChar(state, "l");
    state = processCtrlBackspace(state);
    expect(state.words[0].typed).toBe("");
    expect(state.words[0].chars).toHaveLength(0);
    expect(state.cursorPosition).toBe(0);
  });

  it("does nothing on empty word", () => {
    const state = createInputState(["hello"]);
    const next = processCtrlBackspace(state);
    expect(next.words[0].typed).toBe("");
    expect(next.cursorPosition).toBe(0);
  });

  it("does nothing when finished", () => {
    let state = createInputState(["a"]);
    state = processChar(state, "a");
    state = processSpace(state);
    expect(state.finished).toBe(true);
    const next = processCtrlBackspace(state);
    expect(next.finished).toBe(true);
  });

  it("clears extra characters too", () => {
    let state = createInputState(["hi"]);
    state = processChar(state, "h");
    state = processChar(state, "i");
    state = processChar(state, "x"); // extra
    state = processChar(state, "y"); // extra
    state = processCtrlBackspace(state);
    expect(state.words[0].typed).toBe("");
    expect(state.words[0].chars).toHaveLength(0);
  });
});

describe("processKeystroke", () => {
  it("delegates backspace correctly", () => {
    let state = createInputState(["hi"]);
    state = processKeystroke(state, "h", {});
    state = processKeystroke(state, "", { backspace: true });
    expect(state.words[0].typed).toBe("");
  });

  it("delegates ctrl+backspace correctly", () => {
    let state = createInputState(["hi"]);
    state = processKeystroke(state, "h", {});
    state = processKeystroke(state, "i", {});
    state = processKeystroke(state, "", { backspace: true, ctrl: true });
    expect(state.words[0].typed).toBe("");
  });

  it("delegates space correctly", () => {
    let state = createInputState(["a", "b"]);
    state = processKeystroke(state, "a", {});
    state = processKeystroke(state, " ", {});
    expect(state.currentWordIndex).toBe(1);
  });

  it("ignores control characters", () => {
    const state = createInputState(["hi"]);
    const next = processKeystroke(state, "\x01", {}); // ctrl+a
    expect(next.words[0].typed).toBe("");
  });

  it("ignores empty string input", () => {
    const state = createInputState(["hi"]);
    const next = processKeystroke(state, "", {});
    expect(next.words[0].typed).toBe("");
  });

  it("ignores multi-char input", () => {
    const state = createInputState(["hi"]);
    const next = processKeystroke(state, "abc", {});
    expect(next.words[0].typed).toBe("");
  });
});

describe("createInputState", () => {
  it("initializes keypress counters at zero", () => {
    const state = createInputState(["hello"]);
    expect(state.keypressCorrect).toBe(0);
    expect(state.keypressIncorrect).toBe(0);
  });

  it("handles single-character words", () => {
    const state = createInputState(["a", "b", "c"]);
    expect(state.words).toHaveLength(3);
    expect(state.words[0].word).toBe("a");
  });
});

describe("keypress counters", () => {
  it("increments correct counter for correct chars", () => {
    let state = createInputState(["hi"]);
    state = processChar(state, "h");
    expect(state.keypressCorrect).toBe(1);
    expect(state.keypressIncorrect).toBe(0);
  });

  it("increments incorrect counter for wrong chars", () => {
    let state = createInputState(["hi"]);
    state = processChar(state, "x");
    expect(state.keypressCorrect).toBe(0);
    expect(state.keypressIncorrect).toBe(1);
  });

  it("increments correct counter for correct space", () => {
    let state = createInputState(["hi", "ok"]);
    state = processChar(state, "h");
    state = processChar(state, "i");
    state = processSpace(state);
    // 2 correct chars + 1 correct space = 3
    expect(state.keypressCorrect).toBe(3);
    expect(state.keypressIncorrect).toBe(0);
  });

  it("increments incorrect counter for space on wrong word", () => {
    let state = createInputState(["hi", "ok"]);
    state = processChar(state, "x");
    state = processChar(state, "i");
    state = processSpace(state);
    // x=incorrect, i=correct, space=incorrect (word wrong)
    expect(state.keypressCorrect).toBe(1);
    expect(state.keypressIncorrect).toBe(2);
  });

  it("does not decrement counters on backspace", () => {
    let state = createInputState(["hi"]);
    state = processChar(state, "x"); // incorrect
    expect(state.keypressIncorrect).toBe(1);
    state = processBackspace(state);
    // Counter should stay at 1
    expect(state.keypressIncorrect).toBe(1);
    state = processChar(state, "h"); // correct
    expect(state.keypressCorrect).toBe(1);
    expect(state.keypressIncorrect).toBe(1);
  });
});

describe("processBackspace edge cases", () => {
  it("does nothing when finished", () => {
    let state = createInputState(["a"]);
    state = processChar(state, "a");
    state = processSpace(state);
    expect(state.finished).toBe(true);
    const next = processBackspace(state);
    expect(next.finished).toBe(true);
  });

  it("handles multiple backspaces in a row", () => {
    let state = createInputState(["hello"]);
    state = processChar(state, "h");
    state = processChar(state, "e");
    state = processChar(state, "l");
    state = processBackspace(state);
    state = processBackspace(state);
    state = processBackspace(state);
    expect(state.words[0].typed).toBe("");
    expect(state.cursorPosition).toBe(0);
    // One more backspace at start should do nothing
    state = processBackspace(state);
    expect(state.cursorPosition).toBe(0);
  });

  it("removes extra characters with backspace", () => {
    let state = createInputState(["hi"]);
    state = processChar(state, "h");
    state = processChar(state, "i");
    state = processChar(state, "x"); // extra
    state = processBackspace(state);
    expect(state.words[0].typed).toBe("hi");
    expect(state.words[0].chars).toHaveLength(2);
  });
});

describe("processSpace edge cases", () => {
  it("does nothing when finished", () => {
    let state = createInputState(["a"]);
    state = processChar(state, "a");
    state = processSpace(state);
    expect(state.finished).toBe(true);
    const next = processSpace(state);
    expect(next.finished).toBe(true);
  });

  it("handles word with only wrong characters", () => {
    let state = createInputState(["hi", "ok"]);
    state = processChar(state, "x");
    state = processChar(state, "y");
    state = processSpace(state);
    expect(state.words[0].completed).toBe(true);
    expect(state.currentWordIndex).toBe(1);
  });
});

File: tests/input_processor_test.rs
use buddytype::engine::input_processor::*;

// ── createInputState ────────────────────────────────────────────

#[test]
fn initializes_with_correct_defaults() {
    let state = create_input_state(&["hello".into(), "world".into()]);
    assert_eq!(state.words.len(), 2);
    assert_eq!(state.current_word_index, 0);
    assert_eq!(state.cursor_position, 0);
    assert!(!state.finished);
    assert_eq!(state.words[0].word, "hello");
    assert_eq!(state.words[0].typed, "");
    assert!(!state.words[0].completed);
    assert_eq!(state.historical_error_chars, 0);
}

#[test]
fn initializes_keypress_counters_at_zero() {
    let state = create_input_state(&["hello".into()]);
    assert_eq!(state.keypress_correct, 0);
    assert_eq!(state.keypress_incorrect, 0);
}

#[test]
fn handles_single_character_words() {
    let state = create_input_state(&["a".into(), "b".into(), "c".into()]);
    assert_eq!(state.words.len(), 3);
    assert_eq!(state.words[0].word, "a");
}

// ── processChar ─────────────────────────────────────────────────

#[test]
fn adds_a_correct_character() {
    let state = create_input_state(&["hello".into()]);
    let next = process_char(state, 'h');
    assert_eq!(next.words[0].typed, "h");
    assert_eq!(next.words[0].chars.len(), 1);
    assert!(next.words[0].chars[0].correct);
    assert!(!next.words[0].chars[0].extra);
    assert_eq!(next.cursor_position, 1);
}

#[test]
fn adds_an_incorrect_character() {
    let state = create_input_state(&["hello".into()]);
    let next = process_char(state, 'x');
    assert_eq!(next.words[0].typed, "x");
    assert!(!next.words[0].chars[0].correct);
    assert!(!next.words[0].chars[0].extra);
    assert_eq!(next.historical_error_chars, 1);
}

#[test]
fn marks_extra_characters_beyond_word_length() {
    let state = create_input_state(&["hi".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'i');
    let state = process_char(state, 'x'); // extra
    assert!(state.words[0].chars[2].extra);
    assert!(!state.words[0].chars[2].correct);
    assert_eq!(state.cursor_position, 3);
    assert_eq!(state.historical_error_chars, 1);
}

#[test]
fn does_nothing_when_finished() {
    let mut state = create_input_state(&["a".into()]);
    state.finished = true;
    let next = process_char(state, 'a');
    assert_eq!(next.words[0].typed, "");
}

// ── processSpace ────────────────────────────────────────────────

#[test]
fn advances_to_next_word() {
    let state = create_input_state(&["hello".into(), "world".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'e');
    let state = process_char(state, 'l');
    let state = process_char(state, 'l');
    let state = process_char(state, 'o');
    let state = process_space(state);
    assert!(state.words[0].completed);
    assert_eq!(state.current_word_index, 1);
    assert_eq!(state.cursor_position, 0);
}

#[test]
fn does_not_advance_if_nothing_typed() {
    let state = create_input_state(&["hello".into(), "world".into()]);
    let next = process_space(state);
    assert_eq!(next.current_word_index, 0);
}

#[test]
fn sets_finished_on_last_word() {
    let state = create_input_state(&["hi".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'i');
    let state = process_space(state);
    assert!(state.finished);
    assert!(state.words[0].completed);
}

#[test]
fn allows_advancing_with_incomplete_word() {
    let state = create_input_state(&["hello".into(), "world".into()]);
    let state = process_char(state, 'h');
    let state = process_space(state);
    assert!(state.words[0].completed);
    assert_eq!(state.current_word_index, 1);
}

#[test]
fn space_does_nothing_when_finished() {
    let state = create_input_state(&["a".into()]);
    let state = process_char(state, 'a');
    let state = process_space(state);
    assert!(state.finished);
    let next = process_space(state);
    assert!(next.finished);
}

#[test]
fn handles_word_with_only_wrong_characters() {
    let state = create_input_state(&["hi".into(), "ok".into()]);
    let state = process_char(state, 'x');
    let state = process_char(state, 'y');
    let state = process_space(state);
    assert!(state.words[0].completed);
    assert_eq!(state.current_word_index, 1);
}

// ── processBackspace ────────────────────────────────────────────

#[test]
fn removes_last_character() {
    let state = create_input_state(&["hello".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'e');
    let state = process_backspace(state);
    assert_eq!(state.words[0].typed, "h");
    assert_eq!(state.words[0].chars.len(), 1);
    assert_eq!(state.cursor_position, 1);
}

#[test]
fn goes_back_to_previous_completed_word() {
    let state = create_input_state(&["hi".into(), "there".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'i');
    let state = process_space(state);
    assert_eq!(state.current_word_index, 1);

    let state = process_backspace(state);
    assert_eq!(state.current_word_index, 0);
    assert!(!state.words[0].completed);
    assert_eq!(state.cursor_position, 2);
}

#[test]
fn does_nothing_at_start_of_first_word() {
    let state = create_input_state(&["hello".into()]);
    let next = process_backspace(state);
    assert_eq!(next.words[0].typed, "");
    assert_eq!(next.cursor_position, 0);
}

#[test]
fn keeps_historical_errors_after_deleting_incorrect_char() {
    let state = create_input_state(&["hello".into()]);
    let state = process_char(state, 'x'); // incorrect
    assert_eq!(state.historical_error_chars, 1);

    let state = process_backspace(state);
    assert_eq!(state.words[0].typed, "");
    assert_eq!(state.words[0].chars.len(), 0);
    assert_eq!(state.historical_error_chars, 1);
}

#[test]
fn backspace_does_nothing_when_finished() {
    let state = create_input_state(&["a".into()]);
    let state = process_char(state, 'a');
    let state = process_space(state);
    assert!(state.finished);
    let next = process_backspace(state);
    assert!(next.finished);
}

#[test]
fn handles_multiple_backspaces_in_a_row() {
    let state = create_input_state(&["hello".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'e');
    let state = process_char(state, 'l');
    let state = process_backspace(state);
    let state = process_backspace(state);
    let state = process_backspace(state);
    assert_eq!(state.words[0].typed, "");
    assert_eq!(state.cursor_position, 0);
    // One more at start should do nothing
    let state = process_backspace(state);
    assert_eq!(state.cursor_position, 0);
}

#[test]
fn removes_extra_characters_with_backspace() {
    let state = create_input_state(&["hi".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'i');
    let state = process_char(state, 'x'); // extra
    let state = process_backspace(state);
    assert_eq!(state.words[0].typed, "hi");
    assert_eq!(state.words[0].chars.len(), 2);
}

// ── processCtrlBackspace ────────────────────────────────────────

#[test]
fn ctrl_backspace_clears_entire_word() {
    let state = create_input_state(&["hello".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'e');
    let state = process_char(state, 'l');
    let state = process_ctrl_backspace(state);
    assert_eq!(state.words[0].typed, "");
    assert_eq!(state.words[0].chars.len(), 0);
    assert_eq!(state.cursor_position, 0);
}

#[test]
fn ctrl_backspace_does_nothing_on_empty_word() {
    let state = create_input_state(&["hello".into()]);
    let next = process_ctrl_backspace(state);
    assert_eq!(next.words[0].typed, "");
    assert_eq!(next.cursor_position, 0);
}

#[test]
fn ctrl_backspace_does_nothing_when_finished() {
    let state = create_input_state(&["a".into()]);
    let state = process_char(state, 'a');
    let state = process_space(state);
    assert!(state.finished);
    let next = process_ctrl_backspace(state);
    assert!(next.finished);
}

#[test]
fn ctrl_backspace_clears_extra_characters() {
    let state = create_input_state(&["hi".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'i');
    let state = process_char(state, 'x'); // extra
    let state = process_char(state, 'y'); // extra
    let state = process_ctrl_backspace(state);
    assert_eq!(state.words[0].typed, "");
    assert_eq!(state.words[0].chars.len(), 0);
}

// ── processKeystroke ────────────────────────────────────────────

#[test]
fn delegates_backspace_correctly() {
    let state = create_input_state(&["hi".into()]);
    let state = process_keystroke(state, "h", false, false);
    let state = process_keystroke(state, "", true, false);
    assert_eq!(state.words[0].typed, "");
}

#[test]
fn delegates_ctrl_backspace_correctly() {
    let state = create_input_state(&["hi".into()]);
    let state = process_keystroke(state, "h", false, false);
    let state = process_keystroke(state, "i", false, false);
    let state = process_keystroke(state, "", true, true);
    assert_eq!(state.words[0].typed, "");
}

#[test]
fn delegates_space_correctly() {
    let state = create_input_state(&["a".into(), "b".into()]);
    let state = process_keystroke(state, "a", false, false);
    let state = process_keystroke(state, " ", false, false);
    assert_eq!(state.current_word_index, 1);
}

#[test]
fn ignores_control_characters() {
    let state = create_input_state(&["hi".into()]);
    let next = process_keystroke(state, "\x01", false, false);
    assert_eq!(next.words[0].typed, "");
}

#[test]
fn ignores_empty_string_input() {
    let state = create_input_state(&["hi".into()]);
    let next = process_keystroke(state, "", false, false);
    assert_eq!(next.words[0].typed, "");
}

#[test]
fn ignores_multi_char_input() {
    let state = create_input_state(&["hi".into()]);
    let next = process_keystroke(state, "abc", false, false);
    assert_eq!(next.words[0].typed, "");
}

// ── Keypress counters ───────────────────────────────────────────

#[test]
fn increments_correct_counter_for_correct_chars() {
    let state = create_input_state(&["hi".into()]);
    let state = process_char(state, 'h');
    assert_eq!(state.keypress_correct, 1);
    assert_eq!(state.keypress_incorrect, 0);
}

#[test]
fn increments_incorrect_counter_for_wrong_chars() {
    let state = create_input_state(&["hi".into()]);
    let state = process_char(state, 'x');
    assert_eq!(state.keypress_correct, 0);
    assert_eq!(state.keypress_incorrect, 1);
}

#[test]
fn increments_correct_counter_for_correct_space() {
    let state = create_input_state(&["hi".into(), "ok".into()]);
    let state = process_char(state, 'h');
    let state = process_char(state, 'i');
    let state = process_space(state);
    // 2 correct chars + 1 correct space = 3
    assert_eq!(state.keypress_correct, 3);
    assert_eq!(state.keypress_incorrect, 0);
}

#[test]
fn increments_incorrect_counter_for_space_on_wrong_word() {
    let state = create_input_state(&["hi".into(), "ok".into()]);
    let state = process_char(state, 'x');
    let state = process_char(state, 'i');
    let state = process_space(state);
    // x=incorrect, i=correct, space=incorrect (word wrong)
    assert_eq!(state.keypress_correct, 1);
    assert_eq!(state.keypress_incorrect, 2);
}

#[test]
fn does_not_decrement_counters_on_backspace() {
    let state = create_input_state(&["hi".into()]);
    let state = process_char(state, 'x'); // incorrect
    assert_eq!(state.keypress_incorrect, 1);
    let state = process_backspace(state);
    // Counter should stay at 1
    assert_eq!(state.keypress_incorrect, 1);
    let state = process_char(state, 'h'); // correct
    assert_eq!(state.keypress_correct, 1);
    assert_eq!(state.keypress_incorrect, 1);
}
